### ISSUE TITLE
Prepare for `ppr` and `dynamicIO`

### ISSUE THREAD

**Issue by amannn** [amannn/next-intl#1493]
### Is your feature request related to a problem? Please describe.

Next.js is changing the rendering behavior. It's still very early, and things are moving, but it would be interesting to see if we can adapt `next-intl` to better suit the needs of these rendering modes.

For now, this issue is meant to collect related resources.

**Related**
- https://github.com/amannn/next-intl/issues/1464
- https://github.com/vercel/next.js/discussions/71927
- [something for "global" params](https://x.com/sebmarkbage/status/1851728589240115459)

### Describe the solution you'd like

-

### Describe alternatives you've considered

-

**Comment by amannn** [comment]
Hmm, are we talking about the same thing? [PPR in Next.js](https://nextjs.org/docs/app/api-reference/next-config-js/ppr) refers to _partial prerendering_, but based on your comment I think you mean the same thing?

I haven't had the time to experiment with PPR yet, but am planning to do this in the next few days. I'd be curious if `generateStaticParams` has an effect on whether reading `params` requires dynamic rendering—at least from a user perspective I'd expect it to render fine statically. Not sure if other pieces are relevant though like `'use cache';`.

I'll update this thread here with my findings!

**Comment by zipme** [comment]
Thanks for the update! Yes, I was referring to partial prerendering (I updated my comment). I would have expected `generateStaticParams` to influence this behavior, and it indeed worked in Next 14. However, after upgrading to 15, it no longer functions as expected.

It's still unclear whether `params` qualifies as a dynamic API. It's not listed here: https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-apis, but in the [version 15 upgrade guide](https://nextjs.org/docs/app/building-your-application/upgrading/version-15#async-request-apis-breaking-change), it does mention an async request dynamic API change, and `params` appears in the list.

**Comment by amannn** [comment]
That's interesting, yes. There seems to be a new API coming for accessing top-level params: https://github.com/amannn/next-intl/issues/663#issuecomment-2449100558. Maybe that one will have a different behavior when used with PPR.

**Comment by dBianchii** [comment]
> Thanks for the update! Yes, I was referring to partial prerendering (I updated my comment). I would have expected `generateStaticParams` to influence this behavior, and it indeed worked in Next 14. However, after upgrading to 15, it no longer functions as expected.
> 
> It's still unclear whether `params` qualifies as a dynamic API. It's not listed here: https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-apis, but in the [version 15 upgrade guide](https://nextjs.org/docs/app/building-your-application/upgrading/version-15#async-request-apis-breaking-change), it does mention an async request dynamic API change, and `params` appears in the list.

I believe, anything that is async is considere dynamic under dynamicIO

**Comment by dBianchii** [comment]
would like to see a i18n routing example working with staticly generated pages using dynamicIO

**Comment by TomJD** [comment]
Are there any examples with `dynamicIO`?

I am getting this error:

```
Error: Route "/": A component accessed data, headers, params, searchParams, or a short-lived cache without a Suspense boundary nor a "
use cache" above it. We don't have the exact line number added to error messages yet but you can see which component in the stack belo
w. See more info: https://nextjs.org/docs/messages/next-prerender-missing-suspense
```

as soon as I have this in `RootLayout`:

` const locale = await getLocale();`

it's i18n without routing. I understand it's from the cookie approach, but I can't figure out how to get around it.

```
    "next": "15.1.1-canary.26",
    "next-intl": "4.0.0-beta-40d535a",
```

**Comment by andrewLapshov** [comment]
Hi, @amannn! I'm trying to use dynamicIO in a project along with useCache, but now all pages become dynamic, even with i18 routing. Is this the expected behavior? Using next-intl 4

**Comment by amannn** [comment]
Yes, `rootParams` will be required for static rendering with i18n routing. I found that `setRequestLocale` will not work with `dynamicIO`. It should be easy to adopt based on my experiments, but for the time being I guess we have to wait.

**Comment by MichalMoravik** [comment]
Hey @amannn, I used Next canary to see if I could do PPR and turn most of my page static but `next-intl` was failing and so my attempt wasn't successful. I need PPR to wrap my dynamic Shopping Bag (cart).

I did not use dynamicIO though. Is this issue related to PRR even without dynamicIO on? Do we need `rootParams` in this case too?

Using next-intl 4. Thanks :)

**Comment by amannn** [comment]
Hey @MichalMoravik, if you're using i18n routing, you indeed need `rootParams` for PPR (see my comments above). It's currently available via `unstable_rootParams`, but it seems like the API will change a bit in the near future: https://github.com/vercel/next.js/pull/80255. So it might be best to wait a bit here.

**Comment by MichalMoravik** [comment]
Alright, yea, you're right. Thank you :)

**Comment by JakubFaltyn** [comment]
DynamicIO has been renamed to cacheComponents. Please update the title to avoid confusion :))

**Comment by mtlaso** [comment]
> DynamicIO has been renamed to cacheComponents. Please update the title to avoid confusion :))

From [https://nextjs.org/blog/next-15-4](https://nextjs.org/blog/next-15-4)

> Cache Components (beta): Consolidates experimental caching features (Dynamic IO, use cache, and Partial Prerendering) into a unified cacheComponents flag, simplifying performance optimization strategies. The Next.js 16 blog post will detail our vision.

**Comment by FacundoSpira** [comment]
> For apps without i18n routing, I think `next-intl@4` and the patterns suggested in the docs might work.

Hey @amannn. First of all thanks for the great work! 💪 

I was trying to use next-intl@4 with the latest nextjs@canary version, **without i18n routing**. Docs suggest to do the following in the root layout
```tsx
import {NextIntlClientProvider} from 'next-intl';
import {getLocale} from 'next-intl/server';
 
export default async function RootLayout({
  children
}: {
  children: React.ReactNode;
}) {
  const locale = await getLocale();
 
  return (
    <html lang={locale}>
      <body>
        <NextIntlClientProvider>{children}</NextIntlClientProvider>
      </body>
    </html>
  );
}
```

However, having the `await` triggers a warning/error in Next.js: `A component accessed data, headers, params, searchParams, or a short-lived cache without a Suspense boundary nor a "use cache" above it. See more info: https://nextjs.org/docs/messages/next-prerender-missing-suspense.` In this case I can't really use `rootParams`, since as I mentioned, I'm not using i18n routing. 

Any suggestion?

**Comment by amannn** [comment]
@FacundoSpira Thanks for chiming in here! I guess in your `i18n/request.ts` you have some kind of dynamic function which fetches the user locale, right? That's likely triggering the warning.

Based on https://github.com/vercel/next.js/discussions/71927#discussioncomment-11105573, I think this will be supported in the future:

> It will also be possible to use a Suspense boundary above html but that's not the best solution since it makes the whole page dynamic.

Making the whole page dynamic is essentially what you need with this approach, since the very first piece of your HTML markup (`<html lang="…">`) depends on dynamic user information.

Potentially to get rid of the warning, can you do something like this?

```tsx
export default function RootLayout({children}) {
  return (
    <Suspense fallback={<html lang="en"><body>Loading …</body></html>}>
      <AppLayout />
    </Suspense>
  );
}

async function AppLayout() {
  const locale = await getLocale();
 
  return (
    <html lang={locale}>
      <body>
        <NextIntlClientProvider>{children}</NextIntlClientProvider>
      </body>
    </html>
  );
}
```

I haven't tried it, but maybe worth a shot. Otherwise this might require future improvements in Next.js as mentioned in the linked comment.

But if your whole page relies on request-specific user information, I guess PPR can't prerender too much anyway. I'd guess some strategic `'use cache'` for certain components can be more useful here (esp. for non-user specific parts).

**Comment by LouisCuvelier** [comment]
FYI `rootParams` seems to have been removed : https://github.com/vercel/next.js/pull/84373

**Comment by asgorobets** [comment]
> FYI `rootParams` seems to have been removed : [vercel/next.js#84373](https://github.com/vercel/next.js/pull/84373)

Or likely stabilized, as the unstable_rootParams was removed in favor of compile-time insert, hence they're cleaning up old explicit usage in favor of helper getRootParam: https://github.com/vercel/next.js/blob/4c4a5bc83f690a7248068f05e4d285a4d42e07a0/packages/next/src/build/webpack/loaders/next-root-params-loader.ts#L37

**Comment by amannn** [comment]
Related to using `cacheComponents`, I've opened a discussion over in the Next.js repository to get `dynamicParams = false` working for this mode: https://github.com/vercel/next.js/discussions/84991

If you're interested in this, please leave an upvote on the discussion and feel free to join the conversation!

**Comment by amannn** [comment]
Since `cacheComponents` is available as stable in Next.js 16, the question for support with `next-intl` came up again.

The tldr; is that `cacheComponents` is _not_ supported yet with the capabilities we have today.

**But note**: Next.js 16 by itself (if you don't use `cacheComponents`) is supported as of [`next-intl@4.4`](https://github.com/amannn/next-intl/releases/tag/v4.4.0).

---

**Longer version:**

What we need is [`next/root-params`](https://github.com/vercel/next.js/pull/80255). It's expected to ship in a Next.js 16.x minor version based on the Next.js 16 release blog post.

An experimental version of `next/root-params` is already available in the latest Next.js versions, but in its current state, it can't be used within a `'use cache'` boundary.

The error message you'll see hints at support being added in the future though:

```
Route /[locale] used `import('next/root-params').locale()` inside `"use cache"` or `unstable_cache`.
Support for this API inside cache scopes is planned for a future version of Next.js.
```

So for the time being, it seems like we still have to wait a bit … 

**A bit more on the necessity for `next/root-params`:** 

The mechanism for passing a locale to Server Components that's currently used in `next-intl` is really a [workaround](https://github.com/vercel/next.js/discussions/58862) that glued things together until something like `next/root-params` ships.

If you don't use static rendering, Next.js will complain about a header being read in cached content. And if you use `setRequestLocale`, then you might see cases where the locale is randomly lost. Either isn't a good fit for `cacheComponents`.

The good news is that `next-intl@4` will work with `next/root-params` without a library change being necessary, so everything is already in place. I'll put a blog post out with guidance on how to use it as soon as there's more news regarding this feature! 🙂 

See also: [Reading params deeply in Server Components (e.g. for i18n / multi-tenancy)](https://github.com/vercel/next.js/discussions/58862)

**Temporary workaround:**

If you enable `cacheComponents` but don't use `'use cache'` anywhere, then your app might still work. It will break as soon as you add `'use cache'` somewhere.

However, you can theoretically work around this if you pass an [explicit locale](https://next-intl.dev/docs/usage/configuration#server-request-locale) to APIs like `await getTranslations`.

It might be easy to miss though, so `next/root-params` should really be the right fix here.

**Comment by zipme** [comment]
I'm uncertain whether `next/root-params` will be supported in server actions in addition to 'use cache.' The current `unstable_rootParams` API does support server actions. If root parameters are not supported, will we need to pass the locale to the server action as well?

For example
```tsx
const locale = useLocale()
const someHandler = () => {
  startTransition(() => {
   await someServerAction({ locale })
  })
}
```

**Comment by amannn** [comment]
@zipme Yep, that's still up for debate as far as I know. What makes Server Actions a bit special is that they're put in a queue and by the time they execute, the root param value (or even segment) could be different. I'd certainly really appreciate it if Server Actions were supported.

I hope there can be more public discourse about this once the feature is mentioned in the Next.js docs.

But yeah, the alternative is passing the locale to the action manually and consuming it via `getTranslations({locale})`.

**Comment by ceolinwill** [comment]
Yeah, I noticed passing `setRequestLocale` seems unreliable when using `use cache` with other dynamic params.

For example, I tried `use cache` on a route like this: `/[locale]/posts/[id]`

When I visited that page for the first time, it worked fine: `/en/posts/1`

However, it broke when I changed the locale to `/pt/posts/1`. It was using the previous locale (`en`)

A workaround for this seems to be returning returning at least one locale from `generateStaticParams` for dynamic pages

When I did this, it worked:

```ts
export function generateStaticParams() {
  return [{ id: "1", locale: "en" }];
}
```

I've [created a repo](https://github.com/ceolinwill/next-cache-i18n) where I was playing with cache components and next-intl. It does seem to be working with "use cache"

That being said, despite `cacheComponents` being stable, it still has some weird issues. For example, they [introduced this weird bug](https://github.com/issues/created?issue=vercel%7Cnext.js%7C85218) on canary.17 the night before launching Next.js and making it stable

So, I'd be cautious using `cacheComponents` right now for something in production. I'd wait for 16.1 or 16.2 unless you're starting a new project and can afford some bugs

**Comment by Joshua-hypt** [comment]
Will Next-Intl + cacheComponents work without blocking the route (Blocking Route Error: `Uncached data was accessed outside of <Suspense>`) **without** using locale based routing? Because the suggested fix uses `next/root-params` which reads out the lang parameter.

**Comment by amannn** [comment]
@Joshua-hypt How are you reading the locale in `getRequestConfig`? E.g. if you're reading from cookies, it's naturally a bit hard to cache on the server since it depends on the user request. It seems like something like [`use cache: private`](https://nextjs.org/docs/app/api-reference/directives/use-cache-private) aims to cover cases that include cookies for example, so this could be something to investigate. Note that `use cache: 'private'` is experimental at this point.

If you look into this, please share your experience here!

**Comment by Joshua-hypt** [comment]
@amannn I don't think there is a non-blocking/static way to fetch the locale inside `getRequestConfig` without using locale based routing (correct me if I'm wrong). As far as I understand, `use cache: private` allows for runtime prefetching while using cookies/headers, but on initial render it will still be blocked, requiring a suspense >.<

So since cookies() are async now, it wouldn't be possible?

**Comment by amannn** [comment]
> but on initial render it will still be blocked, requiring a suspense

I think you're right, yes! But it kind of makes sense, right? If you only know the locale of the user once a request hits the server, that's the time you can start rendering with it, right? If you need it to set the `<html lang="…">` attribute, that effectively means that blocking the entire page would be necessary.

However, I guess as your supported locales are a limited set of values, another option could be to use [`localePrefix: 'never'`](https://next-intl.dev/docs/routing/configuration#locale-prefix-never) for this use case, which can help you cache all localized variants while picking the right one with a lightweight ~~middleware~~ proxy that reads the cookie.

Maybe that could be something for your use case?

**Comment by In-line** [comment]
I came up with something like this for my personal needs.

```tsx
import { getTranslations, setRequestLocale } from 'next-intl/server';
import { Suspense } from 'react';
import { Skeleton } from '@mui/material';

export const getTranslationsPromise = async ({
  locale,
  namespace,
}: {  
  locale: Promise<string>;
  namespace: string;
}) => {
  const localeString = await locale;
  const t = await getTranslations({
    locale: localeString,
    namespace,
  });
  setRequestLocale(localeString);
  return t;
};
 
const AsyncTranslatorBlockingImpl = async ({
  translations,
  translationKey,
}: {
  translations: Promise<(key: string) => string>;
  translationKey: string;
}) => {
  const t = await translations;
  return t(translationKey);
};

export const AsyncTranslator = ({translations, translationKey}: {
  translations: Promise<(key: string) => string>,
  translationKey: string,
}) => {
  return (
    <Suspense fallback={<Skeleton variant="text" width={100} height={24} />}>
       <AsyncTranslatorBlockingImpl translations={translations} translationKey={translationKey} />
    </Suspense>
  );
}

export const getTranslationFn = (translations: Promise<(key: string) => string>) => {
  // eslint-disable-next-line react/display-name
  return (translationKey: string) => <AsyncTranslator translations={translations} translationKey={translationKey} />;
}
```

Usage is pretty simple
```tsx
export default async function AboutPage({
  params,
}: {
  params: Promise<{ locale: string }>;
}) {
  const locale = params.then(({ locale }) => locale);
  const translations = getTranslationsPromise({ locale, namespace: 'About' });
  const t = getTranslationFn(translations);

return (
  <Button
       .....
  >
    {t('telegram')}
  </Button>
)
```

The whole route is still kinda blocked on server answer. As I had to put `NextIntlClientProvider` inside a Suspense, but the page is much more responsive. I tested this on NextJS 15.5.1-canary.39 (with cacheComponents = true, reactCompiler = true), because the 16.0.0 has some issues with [Material UI](https://github.com/mui/material-ui/issues/47109).

It's okay, because it's a small scale personal project.

**Comment by fairjm** [comment]
To work with cacheComponent, my current approach is to convert everything into async components (because use cache can only be used in async functions), then get the locale from params and set it using setRequestLocale. Otherwise, all related components afterward would request the cookie(or header?), which means they all need to be wrapped with Suspense.

Places where Link(from next-intl) is used also need to be wrapped with Suspense, preferably with an additional wrapper.

**Comment by Joshua-hypt** [comment]
This might be helpful 

https://github.com/aurorascharff/next-intl-cache-components

**Comment by ceolinwill** [comment]
Yeah, we can use `next-intl` with `use cache` as long as we do this (I'm linking a simple example for each item):

- [Call setRequestLocale](https://github.com/ceolinwill/next-cache-i18n/blob/main/app/%5Blocale%5D/posts/%5Bid%5D/page.tsx#L9)
- Use a [loading file](https://github.com/ceolinwill/next-cache-i18n/blob/main/app/%5Blocale%5D/posts/%5Bid%5D/loading.tsx) OR [wrap it in Suspense](https://github.com/ceolinwill/next-cache-i18n/blob/main/app/%5Blocale%5D/layout.tsx#L49-L51)

**Comment by amannn** [comment]
@aurorascharff wrote a great article that goes further into the current compatibility of `cacheComponents` and `next-intl`: 

→ [**Implementing Next.js 16 'use cache' with next-intl Internationalization**](https://aurorascharff.no/posts/implementing-nextjs-16-use-cache-with-next-intl-internationalization/)

It mentiones a workaround for those desparate in need for using `cacheComponents` today. I've also updated [my comment above](https://github.com/amannn/next-intl/issues/1493#issuecomment-3435882292) to include it as well.

**Comment by dimitrov93** [comment]
Does someone have domain-based route example? I cannot make it work properly :(

**Comment by jjtimi888** [comment]
Does any solution for enable "cacheComponents" and still using get cookie from `i18n/request.ts`

**Comment by nghiepdev** [comment]
> Does any solution for enable "cacheComponents" and still using get cookie from `i18n/request.ts`

As mentioned above regarding ‎`use cache: private`, I see this directive is available on Next.js 16.0.3 stable. For me, enabling ‎`cacheComponents` still causes several issues beyond ‎`next-intl`.
Other issues like `usePathname` and `useSelectedLayoutSegments` also require wrapping with `Suspense`. Too many `Suspense` boundaries create annoying blank loading gaps.

**Comment by markustripp** [comment]
As a workaround you can wrap the `<NextIntlClientProvider>` in a suspense boundary. 

layout.tsx:

```tsx
import { NextIntlClientProvider } from 'next-intl'
import { getUserLocale } from '@/i18n/locale'
import { Suspense } from 'react'

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode
}>) {
  return (
    <html>
      <body>
        <Suspense fallback={null}>
          <NextIntlClientProviderServer>
            {children}
          </NextIntlClientProviderServer>
        </Suspense>
      </body>
    </html>
  )
}

async function NextIntlClientProviderServer({
  children,
}: {
  children: React.ReactNode
}) {
  const locale = await getUserLocale()
  return (
    <NextIntlClientProvider locale={locale}>
      {children}
    </NextIntlClientProvider>
  )
}
```

You can move the `NextIntlClientProviderServer` into a separate file, but I put it directly in the layout.tsx here. `getUserLocale()` gets the current locale from the cookie (see: `locale.ts` below) and passes the value to the `NextIntlClientProvider`. I use it for my [Headcode CMS](https://headcodecms.com/) project and it works fine.

request.ts:

```ts
import { getRequestConfig } from 'next-intl/server'
import { getUserLocale } from './locale'

export default getRequestConfig(async () => {
  const locale = await getUserLocale()

  return {
    locale,
    messages: (await import(`./messages/${locale}.json`)).default,
  }
})
```

locale.ts:

```ts
import { cookies } from 'next/headers'

const COOKIE_NAME = 'next-intl-locale'

export async function getUserLocale() {
  const cookieStore = await cookies()
  return cookieStore.get(COOKIE_NAME)?.value || 'en'
}

export async function setUserLocale(locale: string) {
  const cookieStore = await cookies()
  cookieStore.set(COOKIE_NAME, locale)
}
```

**Comment by Maczuga** [comment]
https://github.com/vercel/next.js/releases/tag/v16.1.0
16.1 stable just dropped.
Can't see the next/root-params mentioned anywhere in changes.

Does that mean we gotta wait for another one? Or simply not mentioned?

**Comment by amannn** [comment]
@Maczuga I think it'd be mentioned in the release notes and I also didn't observe any recent PRs, so I think we'll have to wait a bit more for now.

**Comment by Balthasar42** [comment]
I think in the next minor release will be an feature to use the root params.

> unstable_rootParams
>
> The unstable_rootParams function has been removed. We are working on an alternative API that we will ship in an upcoming minor release.
> 


https://nextjs.org/docs/app/guides/upgrading/version-16#unstable_rootparams

**Comment by JakubFaltyn** [comment]
Anybody find out anything new on this?

**Comment by ihardlight** [comment]
@JakubFaltyn, you can enable cache components and use `<NextIntlClientProvider />` in your root layout at [locale]/layout.tsx. However, you should write `setRequest(locale)` in every page and layout where you want to use `getTranslations()`. Additionally, you should not use the `getLocale()` method or other methods that require the request context. For `getTranslations()` and `useTranslations()` the code will work correctly.

P.S. you can not use getTranslations or other next-int functional in loading.tsx because it does not recieve any params and does not inherit context from layout. If you need for any reason translated text in loading state you can use Suspense in your page.tsx file instead
```react
<Suspense fallback={<LoadingComponent />}>
   <PageComponent />
</Suspense>
```

**Comment by alvarogomezj** [comment]
@ihardlight so if we use `getLocale()` or any other utility in our code that is not supported... what is the alternative in order to use cache components? prop drilling from the page? 

In any case, I suppose this is just a temporal solution

**Comment by ihardlight** [comment]
@alvarogomezj For `page.tsx` and `layout.tsx`, you can always obtain the locale from the parameters using `const { locale } = await params`. This will not affect the static generation or cache components if you write `setRequestLocale(locale)` with it.

For other cases where you want to use the locale directly in a component, you can make it dynamic and wrap it in `Suspense` or use prop drilling. Alternatively, you can create a client component and use the locale from the context. The choice is yours.

These recommendations are based on my experience with next-intl in my projects and may differ from the advice of library maintainers.

**Comment by JakubFaltyn** [comment]
Since the Next.js team doesn’t really seem to care about making `rootParams` stable, I ended up using an undocumented experimental flag with help from Codex. It has been running in production for about two weeks without any issues. 
I’d recommend others here do the same, because after hours of trying different approaches, this was the only thing that actually worked.

In `next.config.ts`:

```ts
experimental: {
  rootParams: true,
},
```

Then use this in your root `layout.tsx`:

```ts
const locale = await getRootLocale();
```

And this is `get-root-locale.tsx`:

```ts
import { notFound } from "next/navigation";
import * as rootParams from "next/root-params";
import { hasLocale } from "next-intl";

import { routing } from "./routing";

const getLocaleFromRootParams = async (): Promise<string> => {
  const localeGetter = rootParams as unknown as { locale: () => Promise<string> };
  return localeGetter.locale();
};

export async function getRootLocale() {
  const locale = await getLocaleFromRootParams();

  if (!hasLocale(routing.locales, locale)) {
    notFound();
  }

  return locale;
}

export async function getRootLocaleOrDefault() {
  const locale = await getLocaleFromRootParams();

  return hasLocale(routing.locales, locale) ? locale : routing.defaultLocale;
}
```

I still can't believe that they pushed cacheComponents to stable without so important feature.

**Comment by wottpal** [comment]
@JakubFaltyn Thanks for sharing! Are you using that in Next.js `stable` or v16.2.0 `canary`?

**Comment by JakubFaltyn** [comment]
@wottpal Currently I'm on stable 16.1.6. Experimental flag for rootParams already available for that version.

**Comment by GrahamQuan** [comment]
its comming https://github.com/vercel/next.js/pull/91191#issue-4057623324

**Comment by alvarogomezj** [comment]
It's now [merged!](https://github.com/vercel/next.js/pull/91191#issue-4057623324) Can't really wait to test a stable / non hacky solution!

**Comment by nsd0dsawazky** [comment]
Next.js 16.2.0 is out now 🎉

**Comment by masterjanic** [comment]
@amannn I am ready lets get the hackathon started

**Comment by amannn** [comment]
Seems like the PR is included with Next.js 16.2, yes! If someone gives this a spin, I’d appreciate it if you could share your experiences here!

My son was born this week, I’m taking a bit of time off to support the family 😊

**Comment by masterjanic** [comment]
https://github.com/virtbase/virtbase/commit/060fef3672dac79b84a2aacbbf352b29db8cb7e9

I tried so far using your old (drafted) blog posting. I have split routing with public pages using rootParams and app using cookies for now. Seems to work without issues. But I had to add a `loading.tsx` file with `null` at the root otherwise it would not let me render my app pages because the `getLocale` is using `cookies()`.

I also had to reorganise my layouts since the root layout exported the viewport and default meta, now split into two layouts.

But I would still say a good cleanup!

**Comment by aurorascharff** [comment]
Congrats on the little one @amannn! 🎉

@masterjanic Nice work getting it running! I also gave it a spin with [#91191](https://github.com/vercel/next.js/pull/91191) in 16.2, and root params work inside `"use cache"` now. [Updated my demo repo](https://github.com/aurorascharff/next-intl-cache-components/commit/8170b93e879cc28596a237e08bee3d2b7e2551e0) and [blog post](https://aurorascharff.no/posts/implementing-nextjs-16-use-cache-with-next-intl-internationalization/#update-nextroot-params-is-here-nextjs-162) so cached components read locale directly via `await rootLocale()` instead of prop-drilling.

About the `loading.tsx` with null workaround, I think that's only needed because `getLocale()` reads from `cookies()`. If you switch to root params for the locale instead, you shouldn't need it since it comes from the URL segment and not from request data. Worth looking into, because a root-level `loading.tsx` means your whole page gets wrapped in a Suspense boundary, which pushes everything out of the static shell. That kind of defeats the purpose of `cacheComponents` since your cached components won't be part of the initial prerender anymore.

One thing I noticed is that you still need to pass locale explicitly to `getTranslations({locale, namespace})` inside `"use cache"`, because `getTranslations('Namespace')` still uses `headers()` under the hood. Would be great if `next-intl` could use root params internally so that's not needed anymore. I think that's been the plan all along.

**Comment by amannn** [comment]
Hey @aurorascharff, I just had a minute to look at your example. You can in fact already avoid reading from headers, please see my PR here: https://github.com/aurorascharff/next-intl-cache-components/pull/6.

Really looking forward to `next/root-params` being released by the Next.js team, I'll share all relevant upgrade hints for `next-intl` users then!

**Comment by aurorascharff** [comment]
Thank you so much @amannn! I've merged the PR and updated the blog post with the new approach. Looking forward to the full next-intl integration once next/root-params is released!

**Comment by wottpal** [comment]
Wow, this PR is impressive, removing so much boilerplate! 🤯 👏

Working flawlessly, only issue I get is caused by this `notFound()`:

```
export default getRequestConfig(async ({ locale }) => {
  if (!locale) {
    const paramValue = await rootParams.locale()
    if (hasLocale(routing.locales, paramValue)) {
      locale = paramValue
    } else {
      notFound() // ⬅️
    }
  }

// …
```

<img width="1962" height="2048" alt="Image" src="https://github.com/user-attachments/assets/a671ca49-ea9d-4251-b56d-734e88aa78de" />

***

Switching back to the good ol' `if (!hasLocale(i18nConfig.locales, locale)) notFound()` in the layout itself resolved it for now! 🙌

**Comment by stefanprobst** [comment]
is `root-params` working in `opengraph-image.tsx` (route handlers) as well?

**Comment by Grod56** [comment]
> Hey [@aurorascharff](https://github.com/aurorascharff), I just had a minute to look at your example. You can in fact already avoid reading from headers, please see my PR here: [aurorascharff/next-intl-cache-components#6](https://github.com/aurorascharff/next-intl-cache-components/pull/6).
> 
> Really looking forward to `next/root-params` being released by the Next.js team, I'll share all relevant upgrade hints for `next-intl` users then!

Is there any workaround to get the new `request.ts` config here to work with the not-found page for non-localized requests? `next/root-params` won't capture `[locale]` when using the configuration prescribed in [your blog post](https://next-intl.dev/docs/environments/error-files#catching-unknown-routes) for handling non-localized requests.

**Comment by zipme** [comment]
Hey, we ran into an issue with `cacheComponents` + `next-intl` on Vercel that might be worth flagging here.

**TL;DR**: With `cacheComponents: true`, any middleware/proxy rewrite on Vercel causes RSC responses to come back with `content-type: text/html` instead of `text/x-component`. The Next.js client router checks this header and falls back to full page reloads. Works fine locally with `next start`.

We use `localePrefix: "never"` with domain-based routing, `rootParams: true` with `next/root-params` to resolve the locale in `getRequestConfig` (instead of relying on `requestLocale` & `setLocale`), and `"use cache"` + `cacheLife` on pages/components. Everything works great locally — the middleware needs to rewrite `/sale/` → `/nl-nl/sale/` to resolve the `[locale]` segment, and that rewrite is what breaks it on Vercel.

We did a bunch of testing on the same Vercel deployment to narrow it down:

| Setup                                   | Rewrite? | Content-Type       | SPA nav?        |
| --------------------------------------- | -------- | ------------------ | --------------- |
| Pages outside `[locale]`, no middleware | No       | `text/x-component` | **Works**       |
| Proxy with `NextResponse.next()` only   | No       | `text/x-component` | 404 (no locale) |
| next-intl middleware (normal flow)      | Yes      | `text/html`        | **Broken**      |
| Manual `NextResponse.rewrite()`         | Yes      | `text/html`        | **Broken**      |
| `beforeFiles` rewrite in next.config.js | Yes      | `text/html`        | **Broken**      |

So it's not next-intl specific — **any rewrite** (middleware or config-based) on Vercel breaks the RSC content-type when `cacheComponents` is enabled.

The RSC body is actually correct (valid Flight payload), and `x-matched-path` shows `/[locale]/[...slug].rsc` — so the server knows it's an RSC request. It's just the `content-type` header that's wrong.

A few more details:

- `x-vercel-cache: HIT` with the same `age` for both RSC and HTML requests, suggesting the CDN isn't properly differentiating variants
- `vary: rsc, next-router-state-tree, ...` is present on the response
- Locally with `next start`, the same rewrite returns `text/x-component` correctly

**Environment**: Next.js 16.2.1, next-intl 4.9+, `cacheComponents: true`, `rootParams: true`, deployed on Vercel.

We couldn't find a workaround that keeps both `cacheComponents` and `localePrefix: "never"`. Disabling `cacheComponents` or switching to `localePrefix: "always"` (no rewrite needed) are the only options we found.

**Comment by icyJoseph** [comment]
@zipme do you happen to know which Vercel CLI version is being used in your builds? I do believe a similar issue was recently fixed, but perhaps there's some edge case still? or you got a Vercel CLI version pinned? People set VERCEL_CLI_VERSION to a given value.

**Comment by zipme** [comment]
> @zipme do you happen to know which Vercel CLI version is being used in your builds? I do believe a similar issue was recently fixed, but perhaps there's some edge case still? or you got a Vercel CLI version pinned? People set VERCEL_CLI_VERSION to a given value.

This: Vercel CLI 50.42.0

**Comment by ignaciogiri** [comment]
@amannn Any workaround if we want to use `localePrefix: "never"`?

**Comment by dimitrov93** [comment]
Whats the exact point of switching from params to root-params here, only to remove setRequestLocale? I'm still getting errors in the server actions and need to do a separate handler to get the locale from the headers, instead of going through getLocale as its going through getRequestConfig(). Am I missing something? Does anyone have working example with server actions? Thanks in advance!

**Comment by zipme** [comment]
> Hey, we ran into an issue with `cacheComponents` + `next-intl` on Vercel that might be worth flagging here.
> 
> **TL;DR**: With `cacheComponents: true`, any middleware/proxy rewrite on Vercel causes RSC responses to come back with `content-type: text/html` instead of `text/x-component`. The Next.js client router checks this header and falls back to full page reloads. Works fine locally with `next start`.
> 
> We use `localePrefix: "never"` with domain-based routing, `rootParams: true` with `next/root-params` to resolve the locale in `getRequestConfig` (instead of relying on `requestLocale` & `setLocale`), and `"use cache"` + `cacheLife` on pages/components. Everything works great locally — the middleware needs to rewrite `/sale/` → `/nl-nl/sale/` to resolve the `[locale]` segment, and that rewrite is what breaks it on Vercel.
> 
> We did a bunch of testing on the same Vercel deployment to narrow it down:
> 
> Setup	Rewrite?	Content-Type	SPA nav?
> Pages outside `[locale]`, no middleware	No	`text/x-component`	**Works**
> Proxy with `NextResponse.next()` only	No	`text/x-component`	404 (no locale)
> next-intl middleware (normal flow)	Yes	`text/html`	**Broken**
> Manual `NextResponse.rewrite()`	Yes	`text/html`	**Broken**
> `beforeFiles` rewrite in next.config.js	Yes	`text/html`	**Broken**
> So it's not next-intl specific — **any rewrite** (middleware or config-based) on Vercel breaks the RSC content-type when `cacheComponents` is enabled.
> 
> The RSC body is actually correct (valid Flight payload), and `x-matched-path` shows `/[locale]/[...slug].rsc` — so the server knows it's an RSC request. It's just the `content-type` header that's wrong.
> 
> A few more details:
> 
> * `x-vercel-cache: HIT` with the same `age` for both RSC and HTML requests, suggesting the CDN isn't properly differentiating variants
> * `vary: rsc, next-router-state-tree, ...` is present on the response
> * Locally with `next start`, the same rewrite returns `text/x-component` correctly
> 
> **Environment**: Next.js 16.2.1, next-intl 4.9+, `cacheComponents: true`, `rootParams: true`, deployed on Vercel.
> 
> We couldn't find a workaround that keeps both `cacheComponents` and `localePrefix: "never"`. Disabling `cacheComponents` or switching to `localePrefix: "always"` (no rewrite needed) are the only options we found.

btw, this bug was confirmed by Vercel and already fixed in latest canary 🎊 https://github.com/vercel/next.js/pull/91541

**Comment by nghiepdev** [comment]
It's getting really close now:  https://preview.nextjs.org/docs/app/api-reference/functions/next-root-params

**Comment by ceolinwill** [comment]
Now that `next/root-params` addresses much of the server-side Cache Components compatibility, I think a separate client-side issue remains in `BaseLink`.

[`BaseLink`](https://github.com/amannn/next-intl/blob/main/packages/next-intl/src/navigation/shared/BaseLink.tsx#L31-L41) calls `usePathname()` for every link. However, the pathname is only passed to [`syncLocaleCookie`](https://github.com/amannn/next-intl/blob/main/packages/next-intl/src/navigation/shared/syncLocaleCookie.tsx), which returns immediately unless the link explicitly changes locales.

As a result, ordinary same-locale links inside dynamic routes require a Suspense boundary when Cache Components are enabled. This is especially awkward when a link appears inside another Suspense boundary's fallback, because it then requires an additional nested boundary.

Would it make sense to split this behavior internally?

- Ordinary same-locale links could render `NextLink` without calling `usePathname()`.
- Locale-changing links could render a small internal component that owns `usePathname()` and synchronizes the locale cookie.

This could preserve the existing `<Link locale="…">` API, forced locale prefixes, `hrefLang`, prefetch behavior, and cookie synchronization while avoiding unnecessary Suspense boundaries for ordinary links.

**Comment by amannn** [comment]
@ceolinwill Interesting. Yeah, I think you have a point there. Would you be up to submit a PR?

**Comment by ceolinwill** [comment]
> Would you be up to submit a PR?

@amannn sure! I'll look into it tomorrow or Thursday

**Comment by GrahamQuan** [comment]
nextjs 16.3 root-params its here 🫶
https://nextjs.org/docs/app/api-reference/functions/next-root-params

**Comment by amannn** [comment]
> nextjs 16.3 root-params its here 🫶

Yes! 🥹🥹🥹

I'm going to close this issue now since that should be the proper fix for `next-intl` to support `cacheComponents`. If someone still runs into a problem, please file a new issue!

Here's the blog post on `next/root-params`:

→ [**Using `next/root-params` in Next.js 16.3**](https://next-intl.dev/blog/nextjs-root-params)
